### PR TITLE
More robust local CMO script

### PR DESCRIPTION
**Pull request by sthaha**
This patch
  * Adds a `make run-local` target that builds the operator and runs
    hack/local-cmo.sh

  * fixes the old behaviour that overwrote all the existing `spec.overrides` in
    clusterversion/version with just CMO in umanaged state to append CMO
    disable to the existing `overrides`

  * validates if all required tools are installed before modifying the cluster

  * fixes disabled shellcheck lints

Signed-off-by: Sunil Thaha <sthaha@redhat.com>
* [x] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: hack/local-cmo.sh [hunk 1]
```diff
@@ -1,28 +1,103 @@
...
+  return $ret
+}
+
+kc() {
+  kubectl -n openshift-monitoring "$@"
+}
+
+disable_managed_cmo(){
+  # NOTE: we can't kubectl patch the spec.overrides since overrides does not
+  # define the patch strategy so as workaround, we get the entire contents of
```
[the remark this answers is not in the document]
Comment (sthaha) [reply hack/local-cmo.sh:37]: @fpetkovski , I would rephrase it as ..
>  # define the patch strategy, so as *a* workaround, we get ...

File: hack/local-cmo.sh [hunk 2]
```diff
@@ -1,28 +1,103 @@
...
 
-set -eu
 
-# Go to the root of the repo
-cd "$(git rev-parse --show-cdup)"
+validate() {
+  local ret=0
+
+  [[ -x operator ]] || {
+    echo "Cannot find local operator binary. Did you run 'make build' ?"
```
[the remark this answers is not in the document]
Comment (sthaha) [reply hack/local-cmo.sh:14]: Yeah we can do that and if compilation fails and let user deal with the failure.

File: hack/local-cmo.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (fpetkovski) [reply hack/local-cmo.sh:37]: I think in general it is better to stick to short and simple sentences that are easy to read and understand. But I wouldn't mind if you want to keep your suggestion.

File: hack/local-cmo.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulfantom) [reply hack/local-cmo.sh:14]: There are at least 3 Makefile targets which would result in creating a binary:
- `make build` - simple way
- `make operator` - invoked by `make build` and it will try regenerating `assets/`
- `make operator-no-deps` - it will only build operator and not regenerate `assets/`

Since we have those 3 it would be better to just say that binary is missing and not running build for the user.

[the remark this answers is not in the document]
Comment (sthaha) [reply hack/local-cmo.sh]: done :)

File: hack/local-cmo.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (sthaha) [reply hack/local-cmo.sh]: I agree with you. Will keep that in mind next time 😇

File: hack/local-cmo.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (fpetkovski) [reply hack/local-cmo.sh]: @paulfantom I am not sure if the three different make targets change things that much. `make build` will be run only if the operator binary does not exist, so the user can still build a custom one. Also, the message tells the user to run `make build` without indicating that there is some other way to build the binary.

[the remark this answers is not in the document]
Comment (sthaha) [reply hack/local-cmo.sh]: @paulfantom, IMHO, it makes sense to ensure that the operator runs fine and runs with the latest changes. Thus, having `make build` would reduce the edit, run, debug to just 
1. edit the code
2. run the script.

Power users are free to add make changes to script. We could even add a `NO_BUILD=true ./hack/local-cmo.sh`

File: hack/local-cmo.sh [hunk 3]
```diff
@@ -1,28 +1,103 @@
...
 # 3. Try to run CMO locally (CMO needs to be built before starting)
 
-set -eu
 
-# Go to the root of the repo
-cd "$(git rev-parse --show-cdup)"
+validate() {
+  local ret=0
+
+  make build || {
```
Comment (paulfantom) [inline hack/local-cmo.sh:13]: Running `make build` will regenerate everything in `assets/`. This shouldn't be a part of this script.

/hold

Comment (sthaha) [reply hack/local-cmo.sh:13]: Please see: https://github.com/openshift/cluster-monitoring-operator/pull/1181#discussion_r639670188 

--- 
@paulfantom, IMHO, it makes sense to ensure that the operator runs fine and runs with the latest changes. Thus, having make build would reduce the edit, run, debug to just

edit the code
(re)run the script.
E.g.

fd .go$ | entr ./hack/local-cmo.sh
Power users are free to add/make changes to script. We could even add a NO_BUILD=true ./hack/local-cmo.sh

Comment (sthaha) [reply hack/local-cmo.sh:13]: @paulfantom could you please elaborate why assets shouldn't be regenerated?

File: hack/local-cmo.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulfantom) [reply hack/local-cmo.sh]: Bear in mind there are cases during development where regenerating everything in `assets/` can happen after running `make build` and when this isn't something that is desired. This isn't an uncommon pattern. On top of that this script is meant mostly for developers behind CMO and we all here are power users. Considering those I don't see any benefit in complicating script to add `NO_BUILD` feature over running two commands.
If we really want to have something that ties `make build` with this script, then let's do it in Makefile by introducing another target.

File: hack/local-cmo.sh (hunk 3 above)
Comment (paulfantom) [reply hack/local-cmo.sh:13]: There are cases during development where we need to figure out how the stack will behave after applying some change to generated kubernetes manifests stored in `assets/`. It is much easier to change those yaml files manually, run the operator (with those changed files) and later tweak jsonnet, than to start figuring out which part of jsonnet code I need to change just to fix something. To put it in concrete example:
Let's say we got a bug about an incorrect label selector for a ServiceMonitor. It would be easy to:
1) start openshift cluster from cluster-bot
2) amend YAML file
3) build operator
4) run this script
5) amend the file again to add more things
6) DO NOT REBUILD OPERATOR as it is not needed because `assets/` are picked in runtime
7) run script again
8) repeat 5-7 until issue is fixed
9) find how to apply this in jsonnet (at this point cluster is no longer needed)
10) create PR or PRs when fix is also necessary in upstream projects

However if we include `make build` in this script, developer would need to find how to apply the fix in jsonnet first and regenerate `assets/` before every run of this script. Especially consider what happens when jsonnet code is vendored from external library like kube-prometheus.

File: hack/local-cmo.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (sthaha) [reply hack/local-cmo.sh]: @fpetkovski done! I hope it is more readable now

File: hack/local-cmo.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (sthaha) [reply hack/local-cmo.sh]: @paulfantom makes sense. thanks for the explanation. 

I have changed the behaviour of the script back to how it was. It only checks if the `operator` binary exists. 

I have also added a `make run-local` target that `build`s and then runs the `local-cmo.sh` script.

File: hack/local-cmo.sh (hunk 3 above)
Comment (sthaha) [reply hack/local-cmo.sh]: Thanks a lot for taking time to explain 🙇

**Comment by sthaha** [comment]
/hold 

The script fails if overrides is empty

**Comment by sthaha** [comment]
/unhold 

fixed the case where `spec.overrides` is null

File: hack/local-cmo.sh [hunk 4]
```diff
@@ -1,28 +1,111 @@
...
+    echo "ERROR: jq not found. See: https://stedolan.github.io/jq/download/"
+    ret=1
+  }
+
+  return $ret
+}
+
+kc() {
+  kubectl -n openshift-monitoring "$@"
+}
```
Comment (paulfantom) [inline hack/local-cmo.sh:39]: What is the benefit of this wrapper? Considering it is used twice and the script still needs to use `kubectl` directly in other cases.

Comment (sthaha) [reply hack/local-cmo.sh:39]: it avoids having to add `-n openshift-monitorng` (so typos are avoided ;-p) ... but more importantly `kc` is context switch to  indicate that the commands are specific to `openshift-monitoring` namespace as opposed to something that is across the cluster (CVO)

**Review by arajkumar — CHANGES_REQUESTED** [review]

**Comment by simonpasquier** [comment]
This can wait until master reopens for 4.9 (hopefully next week).

File: Makefile [hunk 5]
```diff
@@ -40,6 +40,10 @@ clean:
 # Building #
 ############
 
+.PHONY: run-local
+run-local: build
+	KUBECONFIG=$(KUBECONFIG) ./hack/local-cmo.sh
+
 .PHONY: build
 build: operator
```
[the remark this answers is not in the document]
Comment (arajkumar) [reply Makefile:48]: Also I had to do the following changes in order to make this work on macOS,

```
diff --git a/hack/local-cmo.sh b/hack/local-cmo.sh
index 51f79f2b..98fdd183 100755
--- a/hack/local-cmo.sh
+++ b/hack/local-cmo.sh
@@ -96,10 +96,9 @@ main(){
   gojsontoyaml -yamltojson  < $operator_config |
       jq -r '.data["metrics.yaml"]' > $telemetry_conf

-  local -a images
-  readarray -t images < <(images_from_deployment)
+  local -a images=$(images_from_deployment)

-  run ./operator "${images[@]}" \
+  run ./operator ${images} \
     -assets assets/ \
     -telemetry-config $telemetry_conf \
     -kubeconfig "${KUBECONFIG}" \
```

[the remark this answers is not in the document]
Comment (sthaha) [reply Makefile:48]: @arajkumar thanks for trying this out on OSX ... let me see if I can make this work for both. 

I am surprised that osx doesn't have a the shell builtin `readarray` 🤯

we still want to user array for images .. so `"{images[@]}"` should be used. 
I will try and replace readarray with a `while read...`

[the remark this answers is not in the document]
Comment (sthaha) [reply Makefile:48]: @arajkumar 
> GO_BUILD_RECIPE is set to build the target binary for linux. It would be better to have a separate recipe for local cmo to make the macos users happy 

I agree with you that it is nice to have  but perhaps outside the scope of this patch. 

I see that 84e101dab4f92ca886567f6303d476b74f930edb reverts this functionality that was already in `Makefile`

```diff
-GOOS?=$(shell go env GOOS)
-GOARCH?=$(shell go env GOARCH)
 GO111MODULE?=on
 GOPROXY?=http://proxy.golang.org
 export GO111MODULE
@@ -30,7 +28,7 @@ TOOLING=$(EMBEDMD_BIN) $(GOBINDATA_BIN) $(JB_BIN) $(GOJSONTOYAML) $(JSONNET_BIN)
 JSONNET_SRC=$(shell find ./jsonnet -type f -not -path "./jsonnet/vendor*")
 JSONNET_VENDOR=jsonnet/vendor
·
-GO_BUILD_RECIPE=GOOS=$(GOOS) GOARCH=$(GOARCH) CGO_ENABLED=0 go build --ldflags="-s -X $(GO_PKG)/pkg/operator.Version=$(VERSION)"
+GO_BUILD_RECIPE=GOOS=linux CGO_ENABLED=0 go build --ldflags="-s -X $(GO_PKG)/pkg/operator.Version=$(VERSION)"
·
 
```

[the remark this answers is not in the document]
Comment (sthaha) [reply Makefile:48]: @dgrisonnet Do you know why the above was reverted in https://github.com/openshift/cluster-monitoring-operator/pull/960/files ?

File: Makefile [hunk 6]
```diff
@@ -27,7 +29,7 @@ TOOLING=$(EMBEDMD_BIN) $(GOBINDATA_BIN) $(JB_BIN) $(GOJSONTOYAML) $(JSONNET_BIN)
 JSONNET_SRC=$(shell find ./jsonnet -type f -not -path "./jsonnet/vendor*")
 JSONNET_VENDOR=jsonnet/vendor
 
-GO_BUILD_RECIPE=GOOS=linux CGO_ENABLED=0 go build --ldflags="-s -X $(GO_PKG)/pkg/operator.Version=$(VERSION)"
+GO_BUILD_RECIPE=GOOS=$(GOOS) GOARCH=$(GOARCH) CGO_ENABLED=0 go build --ldflags="-s -X $(GO_PKG)/pkg/operator.Version=$(VERSION)"
```
Comment (sthaha) [inline Makefile:32]: @dgrisonnet @paulfantom @simonpasquier  any thoughts on bringing this back which was deleted (accidentally?? 🤔 )  in 
* https://github.com/openshift/cluster-monitoring-operator/commit/84e101dab4f92ca886567f6303d476b74f930edb
* https://github.com/openshift/cluster-monitoring-operator/pull/960/files

File: hack/local-cmo.sh [hunk 7]
```diff
@@ -1,28 +1,112 @@
...
 
-# shellcheck disable=SC2002
-cat manifests/0000_50_cluster-monitoring-operator_04-config.yaml | gojsontoyaml -yamltojson | jq -r '.data["metrics.yaml"]' > /tmp/telemetry-config.yaml
+  run ./operator "${images[@]}" \
+    -assets assets/ \
+    -telemetry-config $telemetry_conf \
+    -kubeconfig "${KUBECONFIG}" \
+    -namespace=openshift-monitoring \
+    -configmap=cluster-monitoring-config \
+    -logtostderr=true -v=4 2>&1 | tee operator.log
```
Comment (arajkumar) [inline hack/local-cmo.sh:109]: nit:

Please add operator.log into .gitignore.

**Review by arajkumar — APPROVED** [review]
lgtm with a nit.

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulfantom) [reply Makefile:50]: We have a ticket for this that @bison is working on - https://issues.redhat.com/browse/MON-1046

File: Makefile (hunk 6 above)
Comment (paulfantom) [reply Makefile:32]: Seems accidental. Let's add it back.

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (sthaha) [reply Makefile:55]: @paulfantom , I was talking about only 
```diff
-GO_BUILD_RECIPE=GOOS=$(GOOS) GOARCH=$(GOARCH) CGO_ENABLED=0 go build --ldflags="-s -X $(GO_PKG)/pkg/operator.Version=$(VERSION)"
+GO_BUILD_RECIPE=GOOS=linux CGO_ENABLED=0 go build --ldflags="-s -X $(GO_PKG)/pkg/operator.Version=$(VERSION)"
```
which seems unrelated to https://issues.redhat.com/browse/MON-1046 🤔

File: Makefile (hunk 6 above)
Comment (paulfantom) [reply Makefile:37]: I don't have opinion on this, but I am also using Linux, so :shrug: Bear in mind than the same build recipe is used to build a binary that is later put into a container. That container needs to be linux one as even macOS is spawning linux virtual machine to run containers.

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (dgrisonnet) [reply Makefile:55]: That's actually related to https://issues.redhat.com/browse/MON-1046, since I brought this improvement to the Makefile build with the fix of this ticket. The original idea behind this ticket was to move the script to golang which is was I was tasked to do. But, the script was buggy, and since we were blocked by it just before a final freeze, so we decided to revert it instead of fixing the bugs. More info here: https://issues.redhat.com/browse/MON-1046?focusedCommentId=15592931&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15592931. That resulted in the build recipe improvements to also be reverted.

That said, I am totally in favor of adding these env variables again.

File: Makefile (hunk 6 above)
Comment (dgrisonnet) [reply Makefile:37]: Ah indeed, I hadn't thought about the local cmo use case when I changed the build recipe. As Pawel mentioned, it doesn't seem very meaningful to make the build cross-platform since CMO is only meant to be run inside of a container.

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (sthaha) [reply Makefile:55]: @paulfantom @dgrisonnet Thank you for clarifying. I think it is a nice improvement and I have already added that to Makefile :). Thanks for doing the ground work. 

@arajkumar thanks for raising the request :)

File: Makefile (hunk 6 above)
Comment (sthaha) [reply Makefile:37]: This just makes it easier to run local-cmo on Mac. For container builds, it should work as  the script just relies on `go env` which should get the values right 🤞

**Review by simonpasquier — COMMENTED** [review]
/lgtm
